### DISCUSSION TITLE
How to Implement Pulse charging in PyBaMM?

### DISCUSSION THREAD

**Discussion by mlnsai** [pybamm-team/PyBaMM#2843] [Q&A]
Hi Team,

I am trying to implement Pulse charging in PyBaMM. My sample protocol for 100 cycles would be

1. Discharge till 2.8V
#### Pulse charge starts ####
2. Charge at 0.5C for 1 minute
3. Rest for 1 minute
4. Repeat Steps 2 & 3 until 4.2V
#### Pulse charge ends #####
5. Repeat all the steps for 100 cycles.

Is there anyway to implement this in PyBaMM. I cannot think of a way to implement step-4 and combining all steps to form a single cycle. 
@brosaplanella @tinosulzer 
Any leads would be appreciated. Thankyou.

**Comment by RuiheLi** [comment]
Hi, the useful notebook for your question will be [Tutorial 5 - Run experiments](https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/Getting%20Started/Tutorial%205%20-%20Run%20experiments.ipynb). According to your description, all these steps should be fairly simple. One thing worth highlighting may be the use of brackets, which allows you to define steps and repeat cycles. I think it is a bit tricky to implement step-4 without  specifying how many repeats to run, because that requires adding very complicated termination conditions. Even for a cycler in real world, this is pretty difficult. I would suggest to estimate the number of repeats to get to 4.2V, by doing a full 0.5C charge for example (I reckon the number would be around 120 if the nominal capacity in your parameter set is right). then during the pulses, do a bit more, so that even if is already 4.2, the solver should skip the remaining steps and create an empty solution for them. 
To sum up, the recommanded setting will roughly be (may subject to minor format change):

```
(Charge at 0.5C for 1 minute or until 4.2V,
Rest for 1 minute,)*120
```
